src/data/set/Union_lift.lean

```diff
@@ -0,0 +1,235 @@
+/-
+Copyright (c) 2021 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import data.set.lattice
+import order.directed
+/-!
+# Union lift
+This file defines `Union_lift` to lift function defined on each of a collection of sets
+to the Union of those sets.
+
+## Main definitions
+
+* `Union_lift` -  Given a Union of sets `Union S`, define a function on the Union by defining
+it on each component, and proving that it agrees on the intersections.
+* `lift_of_eq_Union` Version of `Union_lift` for a set that is propositionally equal to `Union S`.
+  This is mainly motivated by the need to define functions on directed supremums of e.g. subgroups,
+  where the supremum is propositionally, but not definitionally equal to the Union.
+
+## Main statements
+
+There are proofs of the obvious properties of `Union_lift`, i.e. what it does to elements of
+each of the sets in the `Union`, stated in different ways.
+
+There are also three lemmas about `Union_lift` and three lemmas about `lift_of_eq_Union` intended
+to aid with proving that `Union_lift` of `lift_of_eq_Union` is a homomorphism when defined
+on a Union of substructures. There is one lemma each to show that constants, unary functions,
+or binary functions are preserved. These lemmas are:
+
+*`Union_lift_const`
+*`Union_lift_unary`
+*`Union_lift_binary`
+*`lift_of_eq_Union_const`
+*`lift_of_eq_Union_unary`
+*`lift_of_eq_Union_binary`
+
+## Tags
+
+directed union, directed supremum
+-/
+
+variables {α ι : Type*} {β : Type*}
+
+namespace set
+
+/-- Given a Union of sets `Union S`, define a function on the Union by defining
+it on each component, and proving that it agrees on the intersections. -/
+noncomputable def Union_lift (S : ι → set α)
```

Comment:
I somewhat feel like this should be defined on subtypes not sets; without having checked in detail, my memory is that we usually define things in terms of subtypes first, then provide an alias for the set version where we set `p = (∈ s)`

Reply:
The trouble with that is that I can't take a union of subtypes.

Reply:
Is the subtype corresponding to a Union of sets defeq to the subtype by a supr of predicates?